cpp/dolfinx/fem/Form.h

```diff
@@ -188,44 +219,137 @@ class Form
       bool needs_facet_permutations,
       const std::map<std::shared_ptr<const mesh::Mesh<geometry_type>>,
                      std::span<const std::int32_t>>& entity_maps,
-      std::shared_ptr<const mesh::Mesh<geometry_type>> mesh = nullptr)
-      : _function_spaces(V), _coefficients(coefficients), _constants(constants),
-        _mesh(mesh), _needs_facet_permutations(needs_facet_permutations)
+      std::shared_ptr<const mesh::Mesh<geometry_type>> mesh)
+      : _function_spaces(V), _integrals(std::forward<X>(integrals)),
+        _coefficients(coefficients), _constants(constants), _mesh(mesh),
+        _needs_facet_permutations(needs_facet_permutations)
   {
-    // Extract _mesh from FunctionSpace, and check they are the same
-    if (!_mesh and !V.empty())
-      _mesh = V.front()->mesh();
+    namespace md = MDSPAN_IMPL_STANDARD_NAMESPACE;
     if (!_mesh)
-      throw std::runtime_error("No mesh could be associated with the Form.");
+      throw std::runtime_error("Form Mesh is null.");
 
-    for (auto& space : V)
+    // Check consistency of mesh(es)
     {
-      if (_mesh != space->mesh()
-          and entity_maps.find(space->mesh()) == entity_maps.end())
+      // Integration domain mesh is passed, so check that it is (1)
+      // common for spaces and coefficients (2) or an entity_map is
+      // available
+      for (auto& space : V)
       {
-        throw std::runtime_error(
-            "Incompatible mesh. entity_maps must be provided.");
+        if (auto mesh0 = space->mesh();
+            mesh0 != _mesh and !entity_maps.contains(mesh0))
+        {
+          throw std::runtime_error(
+              "Incompatible mesh. argument entity_maps must be provided.");
+        }
+      }
+      for (auto& c : coefficients)
+      {
+        if (auto mesh0 = c->function_space()->mesh();
+            mesh0 != _mesh and !entity_maps.contains(mesh0))
+        {
+          throw std::runtime_error(
+              "Incompatible mesh. coefficient entity_maps must be provided.");
+        }
       }
     }
 
-    // Store kernels, looping over integrals by domain type (dimension)
-    for (auto&& [domain_type, data] : integrals)
+    for (auto V : _function_spaces)
     {
-      if (!std::ranges::is_sorted(data,
-                                  [](auto& a, auto& b) { return a.id < b.id; }))
+      std::map<std::tuple<IntegralType, int, int>,
+               std::variant<std::vector<std::int32_t>,
+                            std::span<const std::int32_t>>>
+          vdata;
+      if (auto mesh0 = V->mesh(); mesh0 == _mesh)
       {
-        throw std::runtime_error("Integral IDs not sorted");
+        for (auto& [key, integral] : _integrals)
+          vdata.insert({key, std::span(integral.entities)});
       }
+      else
+      {
+        auto it = entity_maps.find(mesh0);
+        if (it == entity_maps.end())
+          throw std::runtime_error("No entity map for the mesh.");
```

Comment:
This was already checked above, so maybe we could remove above check or just `assert` here?